Add parse_semantic_spec

michaelBanerjee opened the pull request:
```
feat: add parse_semantic_spec

Parses a given semantic spec into a SemanticSpec object. Updated a Divider object to include tool and language.

---

* Closes #7 

---

Co-authored-by: Michael Banerjee <michaelbanerjee10@gmail.com>
Co-authored-by: Akshar Patel <aksharpatel1233@gmail.com>
Co-authored-by: Kyle Almeida <almeidakyle03@gmail.com>
```

michaelBanerjee commented:
@StoneyJackson @AksharP5 

Hello, parse_semantic_spec and its tests are ready for review. As a overview, the entire implementation of parse_semantic_spec and it's test files are located in:
```src/plcc/load_spec/parse_spec/parse_semantic_spec```

Also, after Akshar's PR was merged into main I updated my branch. However, the files merged are included in his PR as well. There is a commit called "refactor: split and validate rough_spec before parsing" that is the source of this, and why it shows there are so many files changed. Because of this, as well as parse_dividers being updated in this branch, I updated makeDivider() for load_rough_spec_test.py and split_rough_spec_test.py. It shows that the entire file was changed but in reality it was just makeDivider() and an import (showed in conflicting files). I wasn't sure how this should be handled or if it affects anything so I'm asking about it here. 

For clarity, I'll list the other files that have been implemented or updated:
- parse_dividers.py, parse_rough_test.py, and parse_dividers_test.py was updated by Kyle's commits focusing on updating parse_dividers().
- load_rough_spec.py, split_rough_spec.py, validate_rough_spec.py, and validate_rough_spec_test.py, were the other files added from pulling Akshar's merge.
- All the other files, which are in the parse_spec directory, implement parse_semantic_spec and it's tests.

As discussed in our last meeting, there are separate test files for parse_semantic_spec()'s helper functions, which are parse_codeFragments_test.py and parse_targetLocator_test.py. The implementation of these helper functions are included in parse_semantic_spec.py. In order to run these tests I used the bin/test/units.bash command.

AksharP5 commented on src/plcc/load_spec/load_rough_spec/parse_dividers.py:
```diff
@@ -7,14 +7,26 @@
+            elif not m and not m2:
+                yield Divider(tool=tool, language=language, line=line)
         else:
             yield line
```
This function is becoming too long, which means that it can either be split into separate functions, or you can take the class-based approach, which would leave this method outside of the class so it can still be called from other places, however it would call upon a class method to handle the parsing logic. Also, some variable names can be changed to make it simpler to understand at a quick glance, for example the patterns can be more specific to what they are for, as regex is more complicated to quickly understand. Below is a quick example of how the class-based approach could look:

```
def parse_dividers(lines, tool='Java', language='Java'): # no need to define the pattern here anymore
    parser = DividerParser(tool=tool, language=language)
    return parser.parse(lines)

Class DividerParser:
    # initialize the patterns with specific names, something like this
    dividerPattern = re.compile(r'^%(?:\s.*)?$')
    toolLanguagePattern = re.compile(r'^%\s*(?P<tool>\S+)\s(?P<language>\S+).*$')
    toolOnlyPattern = re.compile(r'^%\s*(?P<tool>\S+)\s*$')
    
    def parse(self, lines):
    # add necessary logic/methods to help parse
```

AksharP5 commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_codeFragments_test.py:
```diff
@@ -0,0 +1,43 @@
+%%%
+block
+%%%
+''')))
```
Since this is a helper function, it can be moved to the bottom of the file, also making a similar one for the lines will help reduce dependencies.

AksharP5 commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_semantic_spec.py:
```diff
@@ -0,0 +1,89 @@
+    targetLocator = None
+    for obj in list_of_Line_and_Block:
+        if isinstance(obj, Line):
+            if re.match(r'\s*#', obj.string) or re.match(r'\s*$', obj.string):
```
the regex matching can be defined in its own function to make the readability of the function simpler, something as simple as:

```
def _isCommentOrBlank(obj_str):
    return re.match(r'\s*#', obj_str) or re.match(r'\s*$', obj_str):
```

which would then make the if statement easier to read

```
if _isCommentOrBlank(obj.string):
```

AksharP5 commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_semantic_spec.py:
```diff
@@ -0,0 +1,89 @@
+    if targetLocator != None:
+        raise CodeFragmentMissingBlockError(targetLocator.line)
+
+    return codeFragmentList
```
This function is very long, which means that it can likely be split up, again a class-based approach can be useful here, similar to what I stated with the parse_dividers. Also, the current variable name list_of_Divider_Line_and_Block is very descriptive, however it may not need to be as long to accomplish the same goal, displaying the type of the object can make it easy for readers to understand, while keeping the name shorter.

```
def parse_semantic_spec(semantic_spec: list[Line | Block | Divider): 
    # instantiate the class and call upon the method, and return the appropriate value

class SemanticSpecParser:
    # all necessary methods to help parse
```

AksharP5 commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_semantic_spec.py:
```diff
@@ -0,0 +1,89 @@
+        name = match.group(1)
+        modifier = match.group(2) if match.group(2) else None
+        return TargetLocator(line = line, className=name, modifier=modifier)
+    raise InvalidTargetLocatorError(line)
```
If you decide to go with the class-based approach, this will likely become a function within the class

AksharP5 commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_semantic_spec_test.py:
```diff
@@ -0,0 +1,54 @@
+block
+%%%
+''')))
+
```
Similar to the other test files, the helper functions can go all the way to the bottom, and there should be others for any object you are using, such as a Line or Divider to reduce dependencies for the tests themselves.

AksharP5 commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_targetLocator_test.py:
```diff
@@ -0,0 +1,25 @@
+from plcc.load_spec.load_rough_spec.parse_lines import Line, parse_lines
+
+def test_ignore_EOL_comments():
+    line = Line('Class:init #comment', 1, None)
```
Similar to above, try not to define objects within the tests themselves, if possible, as it adds a level of dependency that can be handled in its own function, and the test can call upon that function to create necessary objects.

AksharP5 commented:
@StoneyJackson @michaelBanerjee 

I went ahead and reviewed the changes, and left some comments on what I think could be improved, however if anyone has any other ideas that is perfectly fine. Also, @michaelBanerjee for the commit issue, try squashing your commits, and re-commit your changes, I tried it for your branch myself and it seemed to work, however I did not push the changes.

1. reset the commit log to the reorg branch's latest commit, while keeping any changes: ```git reset --soft reorg```
2. ensure that only your new changes are staged for commit, then commit ```git commit -m "message"```

michaelBanerjee commented:
@AksharP5 @StoneyJackson
I have moved the helper functions parse_codeFragments() and parse_targetLocator() into separate files, and made helper a helper class for parse_codeFragements(). Since parse_targetLocator() is a smaller function, I couldn't find many ways to simplify into a helper class so I kept it the same. However keeping these functions in their own files made imports less cluttered and it easier to tell what functions throw errors. With dataclasses also being defined in these files as well, whoever is reading has an isolated view of how they are being constructed which helps with readability.

AksharP5 commented on src/plcc/load_spec/load_rough_spec/parse_dividers_test.py:
```diff
@@ -28,16 +28,33 @@ def test_non_dividers_pass_through():
     lines = list(parse_lines('%'))
     assert list(parse_dividers(lines)) == [
-        Divider(Line('%', 1, None)),
+        Divider(tool='Java', language='Java', line=Line('%', 1, None)),
```
For the line parameter, there is no need to duplicate code and create a new line object, you can simply slice the lines variable, and this should result in the test passing, also creating a helper function to create the divider and block objects for the tests will reduce dependencies, and allow for any future changes to be made in only one place:
```
def test_one_divider():
    lines = list(parse_lines('%'))
    assert list(parse_dividers(lines)) == [
        make_divider(tool='Java', language='Java', line=lines[0]),
    ]

...

def test_blocks_mask_dividers():
    lines = list(parse_blocks(parse_lines('%%%\n%\n%%%')))
    assert list(parse_dividers(lines)) == [
        make_block([
            lines[0].lines[0],
            lines[0].lines[1],
            lines[0].lines[2]
        ])
    ]

def make_divider(tool, language, line):
    return Divider(tool=tool, language=language, line=line)

def make_block(lines):
    return Block(lines)
```

AksharP5 commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_codeFragments.py:
```diff
@@ -0,0 +1,77 @@
+    def parse(self):
+        for obj in self.lines_and_blocks:
+            if isinstance(obj, Divider):
+                raise DividerMustBeginSectionError(obj.line)
```
I am not sure if this should be a check, as each semantic section should only have one Divider and it should always be the first element, keeping the check doesn't hurt, however it seems redundant. The parser should be under the assumption that the first object is the Divider, and all other objects in the section are Line or Block objects, as that is how the split_rough_spec handles the section.

AksharP5 commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_codeFragments.py:
```diff
@@ -0,0 +1,77 @@
+
+    def _parse_line(self, line):
+        if self._isCommentOrBlank(line.string):
+                return # skips blank or comment Lines
```
The comment can be removed, as it is self-explanatory.

AksharP5 commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_codeFragments_test.py:
```diff
@@ -0,0 +1,45 @@
+def test_basic():
+    lines_and_blocks = [codeFragment_line_for_tests, block_for_tests1]
+    assert parse_codeFragments(lines_and_blocks) == [
+        CodeFragment(TargetLocator(lines_and_blocks[0], 'Class', 'init'), block_for_tests1)]
```
Similar to the other tests, creating helper functions to create these objects as opposed to creating them within the tests themselves will reduce dependencies, even if it is for the expected result.

AksharP5 commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_codeFragments_test.py:
```diff
@@ -0,0 +1,45 @@
+codeFragment_line_for_tests = Line('Class:init', 1, None)
+
+block_for_tests1 = make_block()
+block_for_tests2 = make_block()
```
Rather than using variables, keeping these as method calls within the tests will be better, these block variables are unnecessary , and rather than hardcoding a specific Line into a variable, creating a method to create Line objects will be more useful, as you are defining other Line objects above in the tests, if you choose to reuse the same line over and over that can be its own method. This can be applied to all other tests as well.

AksharP5 commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_semantic_spec.py:
```diff
@@ -0,0 +1,21 @@
+
+def parse_semantic_spec(semantic_spec: list[Divider | Line | Block]) -> SemanticSpec:
+    if not isinstance(semantic_spec[0], Divider):
+        raise DividerMustBeginSectionError(semantic_spec[0])
```
Similar to above, I believe this section should be under the assumption that the first element is a Divider, so this check seems redundant.

AksharP5 commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_semantic_spec_test.py:
```diff
@@ -0,0 +1,53 @@
+%%%
+''')))
+
+divider_for_tests = Divider(tool='Java', language='Java', line=Line('%', 1, None))
+
+codeFragment_line_for_tests = Line('Class:init', 1, None)
+
+block_for_tests1 = make_block()
+block_for_tests2 = make_block()
```
Similar to above, having these global variables be methods would be more useful, that way you only define objects such as a Line in one place, as you are still defining a Line object above, even though you have this here. You can have parameters for these to make them more versatile, as the design you currently have requires you to instantiate a new object if it differs slightly from your pre-defined object variables.

AksharP5 commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_targetLocator_test.py:
```diff
@@ -0,0 +1,25 @@
+from plcc.load_spec.load_rough_spec.parse_lines import Line
+
+def test_ignore_EOL_comments():
+    line = Line('Class:init #comment', 1, None)
+    assert parse_targetLocator(line) == TargetLocator(line, 'Class', 'init')
+
+def test_requires_colon():
+    with raises(InvalidTargetLocatorError):
+        targetLocator = parse_targetLocator(Line('Class init', 1, None))
+
+def test_modifier():
+    lines = [Line('Class:init', 1, None),Line('Class:modifier', 1, None)]
+    assert parse_targetLocator(lines[0]) == TargetLocator(lines[0], 'Class', 'init')
+    assert parse_targetLocator(lines[1]) == TargetLocator(lines[1], 'Class', 'modifier')
+
+def test_empty_line_raises_error():
+    with raises(InvalidTargetLocatorError):
+        parse_targetLocator(Line('', 1, None))
+
+def test_className():
+    lines = [Line('Class', 1, None), Line('name', 2, None)]
+    assert parse_targetLocator(lines[0]) == TargetLocator(lines[0], 'Class', None)
+    assert parse_targetLocator(lines[1]) == TargetLocator(lines[1], 'name', None)
```
Once again, the objects within the tests should really be defined in a single method, as opposed to having each test define them. This can be done for all objects that you are instantiating in these tests.

AksharP5 reviewed (changes requested):
@michaelBanerjee 

I just have a few recommendations, mostly regarding the test files, but other than that I believe this looks very good. Please take a look at the comments and let me know if you have any questions!

AksharP5 commented:
@michaelBanerjee @StoneyJackson 

After speaking with @michaelBanerjee we believed that the parse_dividers method could be refactored to improve readability and maintainability for the long-term. Since I already had this done, Michael agreed to push my changes for this, please let me know if you preferred the way it was before as one big method, or if there are any other changes to be made.

StoneyJackson commented:
There are conflicts. Looks minor. But they need to be resolved.

I'll try to take a closer look tomorrow before our meeting (sorry, ran out of steam tonight).

StoneyJackson commented on src/plcc/load_spec/load_rough_spec/parse_dividers.py:
```diff
@@ -7,14 +7,69 @@
-    for line in lines:
-        if isinstance(line, Line) and pattern.match(line.string):
-            yield Divider(line)
+class ParseDividersHelper:
```
Rename to DividerParser.

"Helper" is vague. How does it help? In what way? Once you answer that, put it in the name.

StoneyJackson commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_codeFragments.py:
```diff
@@ -0,0 +1,71 @@
+    targetLocator: TargetLocator
+    block: Block
+
+def parse_codeFragments(lines_and_blocks: list[Line | Block]):
```
Please rename this function and the containing file/module parse_code_fragments.

StoneyJackson commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_codeFragments.py:
```diff
@@ -0,0 +1,71 @@
+    def parse(self):
+        for obj in self.lines_and_blocks:
+
+            handler = {
```
Extract another function or two here...

def parse(self):
    for obj in self.lines_and_blocks:
        self._parse_line_or_block(obj)
    ...

StoneyJackson reviewed (changes requested):
Mostly cosmetic.

StoneyJackson commented on src/plcc/load_spec/load_rough_spec/parse_dividers_test.py:
```diff
@@ -50,9 +67,15 @@ def test_blocks_mask_dividers():
+def make_divider(tool, language, line):
+    return Divider(tool=tool, language=language, line=line)
+
+def make_block(lines):
```
This is unused. So let's delete it.

I figured out it was unused because it showed up in the coverage report.

StoneyJackson commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_code_fragments.py:
```diff
@@ -0,0 +1,74 @@
+                Block: self._parse_block
+            }.get(type(obj))
+
+        if handler:
```
Coverage report claims that no test covers the case when handler is False. Is this really possible? If not, then delete the if. If it is possible, then please create a test that demonstrates it.

Based on the constructor, the type of lines_and_blocks should be a list of Lines or Blocks, and nothing else. So I think this guard is unnecessary. This can only happen if other code in our system incorrectly calls your function. In which case it will correctly explode.

So I think you should delete the if (but keep the body). And implement a test that demonstrates the explosion as a warning to other developers to not pass you a list with something other than a Line or a Block.

StoneyJackson commented on src/plcc/load_spec/parse_spec/parse_semantic_spec/parse_code_fragments.py:
```diff
@@ -0,0 +1,74 @@
+        if self.targetLocator != None:
+            raise DuplicateTargetLocatorError(line.string)
+
+        if re.match(self.targetLocator_regex, line.string):
```
Coverage report claims that no test covers the case when this is false. So as before, can you think of such a case. And is it based on user error or programmer error? If user, add a test. I programmer, remove the guard and add a test to demonstrate how it will fail if missused.

StoneyJackson reviewed (changes requested):
There are some test coverage issues to deal with. It's better that we catch them now. Otherwise they linger and no one fixes them.

Speaking of... I need to raise some of these related to others' that we just merged.

michaelBanerjee committed:
feat: add parse_semantic_spec

Parses a given semantic spec into a SemanticSpec object. Updated a Divider object to include tool and language.

---

* Closes #7

---

Co-authored-by: Michael Banerjee <michaelbanerjee10@gmail.com>
Co-authored-by: Akshar Patel <aksharpatel1233@gmail.com>
Co-authored-by: Kyle Almeida <almeidakyle03@gmail.com>

StoneyJackson reviewed (approved):
Looks good. Thanks!
